include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h

```diff
@@ -152,18 +152,77 @@ __pattern_transform_scan_base_impl(_ExecutionPolicy&& __exec, _Iterator1 __first
     }
 }
 
+// TODO In C++20 we may try to use std::equality_comparable
+template <typename _Iterator1, typename _Iterator2, typename = void>
+struct is_equality_comparable : std::false_type
+{
+};
+
+// Two sycl_iterator's (with different access modes)
+template <sycl::access::mode _Mode1, sycl::access::mode _Mode2, typename _T, typename _Allocator>
+struct is_equality_comparable<sycl_iterator<_Mode1, _T, _Allocator>, sycl_iterator<_Mode2, _T, _Allocator>>
+    : std::true_type
+{
+};
+
+// All with implemented operator ==
 template <typename _Iterator1, typename _Iterator2>
-constexpr bool
+struct is_equality_comparable<_Iterator1, _Iterator2,
+                              std::void_t<decltype(::std::decay_t<_Iterator1>() == ::std::decay_t<_Iterator2>())>>
+    : std::true_type
+{
+};
+
+// Check the ability to compare two sycl_iterator's
+/*
+ * @return bool - true, if these iterators may be the same;
+ *         false - we know these iterators are different
+ */
+template <sycl::access::mode _Mode1, sycl::access::mode _Mode2, typename _T, typename _Allocator>
+bool
+__check_if_iterator_equality_is_possible(sycl_iterator<_Mode1, _T, _Allocator> __it1,
+                                         sycl_iterator<_Mode2, _T, _Allocator> __it2)
+{
+    const auto buf1 = __it1.get_buffer();
+    const auto buf2 = __it2.get_buffer();
+
+    // We are unable to compare two sycl_iterator's from different buffers
+    if (buf1 != buf2)
+        return false;
+
+    // We are unable to compare two sycl_iterator's if one of them is sub_buffer
+    if (buf1.is_sub_buffer() || buf2.is_sub_buffer())
+        return false;
+
+    return true;
+}
+
+template <typename _Iterator1, typename _Iterator2>
+bool
+__check_if_iterator_equality_is_possible(_Iterator1, _Iterator2)
+{
+    return true;
+}
```

Comment:
I think for arbitrary iterators the check should return false, as we cannot assume two instances can be compared.

Reply:
We have 3-level check of two iterators:

1 level - we check in compile time that we are able to compare two iterators:
```
    // 1. operator==(_Iterator1, _Iterator2) should be available in compile-time
    if constexpr (!is_equality_comparable<_Iterator1, _Iterator2>::value)
    {
        return false;
    }
```
2 level - we check some additional requirements for compare two iterators:
 - for sycl_iterator - that they describes the same buffer and so on:
```
template <sycl::access::mode _Mode1, sycl::access::mode _Mode2, typename _T, typename _Allocator>
bool
__check_if_iterator_equality_is_possible(sycl_iterator<_Mode1, _T, _Allocator> __it1,
                                         sycl_iterator<_Mode2, _T, _Allocator> __it2)
{
    // ...
}
```
 - for iterators of all other types there are no requirements:
```
template <typename _Iterator1, typename _Iterator2>
bool
__check_if_iterator_equality_is_possible(_Iterator1, _Iterator2)
{
    return true;
}
```

3 level - we compare two iterators and return result:
```
        // 3. Now we are ready to compare two iterators
        return __it1 == __it2;
```

So up to me the function seems correct:
```
template <typename _Iterator1, typename _Iterator2>
bool
__check_if_iterator_equality_is_possible(_Iterator1, _Iterator2)
{
    return true;
}
```